wire-ios-data-model/Source/Model/UserClient/UserClient+Patches.swift

```diff
@@ -43,23 +75,62 @@ extension UserClient {
         }
     }
 
+    // Problem:
+    // Existing proteus sessions are qualified with user id and client id only. In a
+    // federated environment, there is not the risk of id collisions, since in two
+    // domains there possibly could be users that share the same user id and client id.
+    //
+    // Solution:
+    // Migrate to V3 proteus session ids, which incorporate the domain.
+
     static func migrateAllSessionsClientIdentifiersV3(in moc: NSManagedObjectContext) {
+        WireLogger.proteus.info("migrating all session ids to v3")
+
         let request = UserClient.sortedFetchRequest()
 
-        guard
-            let allClients = moc.fetchOrAssert(request: request) as? [UserClient],
-            let keyStore = moc.zm_cryptKeyStore
-        else {
-            // no client? no migration needed
+        guard let allClients = moc.fetchOrAssert(request: request) as? [UserClient] else {
+            WireLogger.proteus.info("migrating all session ids to v3: no clients to migrate")
             return
         }
 
-        // TODO: [John] use flag here
+        // Important:
+        // If we're migrating from Cryptobox to Core Crypto, it's important that we
+        // migrate the session ids before moving the session files into Core Crypto.
+        //
+        // However, we would only have an instance of `ProteusService` (backed by Core
+        // Crypto) and not an instance of the `UserClientKeyStore` (backed by Cryptobox).
+        //
+        // Migration of the session ids relies on the keystore, so if we don't have an
+        // instance if the keystore already, we create one temporarily for this migration
+        // assuming that we'll perform the migration from Cryptobox to Core Crypto afterwards.
+
+        let keyStore: UserClientKeysStore
+
+        if let existingKeyStore = moc.zm_cryptKeyStore {
+            WireLogger.proteus.info("migrating all session ids to v3: using existing keystore")
+            keyStore = existingKeyStore
+        } else {
+            WireLogger.proteus.info("migrating all session ids to v3: creating temp keystore")
+
+            guard
+                let accountDirectory = moc.accountDirectoryURL,
+                let applicationContainer = moc.applicationContainerURL
+            else {
+                fatalError("Can not migration proteus session ids")
+            }
+
+            keyStore = UserClientKeysStore(
+                accountDirectory: accountDirectory,
+                applicationContainer: applicationContainer
+            )
+        }
+
```

Comment:
This code looks like it duplicates the same code defined around line 50, is this intentional?

Reply:
I will de-duplicate it